External/Notepad/Storage.swift

```diff
@@ -15,9 +15,21 @@
 @objc
 class Storage: NSTextStorage {
 
+    /// Indicates if the SelectionRange is biased (and the UI layer should, instead, consume `overrideSelectionRange`), or not
+    /// See `perserveRealEditedRange` for details.
+    ///
+    @objc
+    private(set) var shouldOverrideSelectionRange = false
+
+    /// Contains the "Real" Edited Range.
+    /// See `perserveRealEditedRange` for details.
+    ///
+    @objc
+    private(set) var overrideSelectionRange = NSRange()
```

Comment:
If `overrideSelectionRange` was an optional we could maybe do away with the boolean `shouldOverrideSelectionRange`.  Just guard for `nil` to know whether to use the default selectedRange or not.

Just a thought.

Reply:
Definitely approached things that way, **but** Objective C. Only way to determine, in ObjC, if a range is invalid, was to use a negative location as a marker, that's why I deferred to the dual property approach.

Like those old ObjC API's... `-(BOOL)doSomethingWithError:(NSError **)error`. I wish we could swift our way out, though!!